sklearn/datasets/tests/test_openml.py

```diff
@@ -0,0 +1,408 @@
+        # single target, so target is vector
+        assert data_by_id.target.shape == (expected_observations, )
+    elif isinstance(target_column_name, list):
+        # multi target, so target is array
+        assert data_by_id.target.shape == (expected_observations,
+                                           len(target_column_name))
+    assert data_by_id.data.dtype == exptected_data_dtype
+    assert data_by_id.target.dtype == exptected_target_dtype
+    assert len(data_by_id.feature_names) == expected_features
+    for feature in data_by_id.feature_names:
+        assert isinstance(feature, string_types)
+
+    if compare_default_target:
+        # check whether the data by id and data by id target are equal
+        data_by_id_default = fetch_openml(data_id=data_id, cache=False)
+        if data_by_id.data.dtype == np.float64:
+            np.testing.assert_allclose(data_by_id.data,
+                                       data_by_id_default.data)
+        else:
+            assert np.array_equal(data_by_id.data, data_by_id_default.data)
+        if data_by_id.target.dtype == np.float64:
+            np.testing.assert_allclose(data_by_id.target,
+                                       data_by_id_default.target)
+        else:
+            assert np.array_equal(data_by_id.target, data_by_id_default.target)
+
+    if expect_sparse:
+        assert isinstance(data_by_id.data, scipy.sparse.csr_matrix)
+    else:
+        assert isinstance(data_by_id.data, np.ndarray)
+
+    # check numeric features. Note that the response of _get_data_features is
+    # mocked too.
+    feature_name_type = {feature['name']: feature['data_type']
+                         for feature in _get_data_features(data_id)}
+    for idx, feature_name in enumerate(data_by_id.feature_names):
+        if feature_name_type[feature_name] == 'numeric':
+            # check that all elements in an object array are numeric
+            # cf. https://stackoverflow.com/a/19486803/1791279
+            assert np.issubdtype(np.array(list(data_by_id.data[:, idx])).dtype,
+                                 np.number)
+
+    return data_by_id
+
+
+def _determine_default_features(data_id, expected_default_target):
+    # fetch features
+    features = _get_data_features(data_id)
+    default_target = _determine_default_target(features)
+    assert expected_default_target == default_target
+
+
+def _monkey_patch_webbased_functions(context, data_id, gziped_files=True):
+    url_prefix_data_description = "https://openml.org/api/v1/json/data/"
+    url_prefix_data_features = "https://openml.org/api/v1/json/data/features/"
+    url_prefix_download_data = "https://openml.org/data/v1/"
+    url_prefix_data_list = "https://openml.org/api/v1/json/data/list/"
+
+    path_suffix = ''
+    read_fn = open
+    if gziped_files:
+        path_suffix = '.gz'
+        read_fn = gzip.open
+
+    def _mock_urlopen_data_description(url):
+        assert (url.startswith(url_prefix_data_description))
```

Comment:
please remove the unnecessary parentheses here and below.

Reply:
Agreed